### PR TITLE
Fix 83: Reload support

### PR DESCRIPTION
**Pull request by delgermurun**
add support for #83

### CODE REVIEW

**Comment by delgermurun** [comment]
Seems like `page.replace` does all job that we need :).

Please check this out @arunoda

**Comment by arunoda** [comment]
Does it rerun triggers and the action? If so we are good to go. 

I think before this we need to work on idempotant routes. We don't rerun the same route even we received it from page. 

Then for the reload, we need to the reload inside a Meteor.EnvironmentVariable. Based on that we can allow to bypass reload. 

I think it's a good idea to implement both in a single PR.

**Comment by delgermurun** [comment]
> Does it rerun triggers and the action? If so we are good to go.

Yes it does.

> I think before this we need to work on idempotant routes. We don't rerun the same route even we received it from page.

Ok, I'll work on it.

> Then for the reload, we need to the reload inside a Meteor.EnvironmentVariable. Based on that we can allow to bypass reload.

Sorry, I don't quite understand what do you mean. Can you explain further?

**Comment by arunoda** [comment]
Did you check this: https://github.com/meteorhacks/flow-router/blob/master/client/router.js#L279

* Let's say we are implementing idempotent logic here: https://github.com/meteorhacks/flow-router/blob/master/client/router.js#L50
* That way we can work on the direct clicks other than coming from `FlowRouter.go()`. 
* Then we will need to face an issue when dealing with `.reload()`
* `Meteor.EnvironmentVariable` is the way to solve this. See: https://github.com/meteorhacks/flow-router/blob/master/client/router.js#L50

**Comment by delgermurun** [comment]
@arunoda thanks, got it.

**Comment by delgermurun** [comment]
@arunoda Pushed partial implementation of  idempotent routes. Missing part is exit triggers.

Do you have any idea about how to bypassing exit triggers?

**Comment by arunoda** [comment]
Can't we do the same Meteor.EnvVar thing inside triggers?

**Comment by delgermurun** [comment]
@arunoda It is impossible (?) to know next path on exit triggers and exit triggers executes before the handler.

I thinks we should do plugin for idempotent routes, probably overriding `page.show`.

**Comment by arunoda** [comment]
Hmm. That's true.
What do you mean by a plugin?
Overriding page.js ?

**Comment by delgermurun** [comment]
Yeah, kind of.

There is [plugins](https://github.com/visionmedia/page.js#available-plugins) that overrides `page.show` method. e.g. [page.js-body-parser.js](https://github.com/kethinov/page.js-body-parser.js/blob/master/page.js-body-parser.js#L127).

**Comment by arunoda** [comment]
Hmm. Okay.

**Comment by delgermurun** [comment]
@arunoda Overrode `page.show`. Please check this out.

File: client/router.js [hunk 1]
```diff
@@ -302,6 +307,18 @@ Router.prototype._notfoundRoute = function(context) {
 Router.prototype.initialize = function() {
   var self = this;
   this._updateCallbacks();
+
+  var oldShow = this._page.show;
+
+  this._page.show = function(path, state, dispatch, push) {
```
Comment (arunoda) [inline client/router.js]: I think we need more details on what's going on here.
So, it's clear on the future too.

**Comment by arunoda** [comment]
This looks great to me.
Let's add some tests and move this into a release.

**Comment by delgermurun** [comment]
Added tests & comment.

PS: My commenting skill isn't good enough. Please add more explanation if you want :).

**Comment by arunoda** [comment]
Great.
You can't ignore that. Practice and learn it.

**Comment by arunoda** [comment]
In the comment mention why you need to override. 
Actually, we all know it's been overridden. What we need to know why?

Why can't we do it without overriding. Answer that and update the PR. 

You need to learn to do that :)

**Comment by arunoda** [comment]
We will also need a test checking triggers as well.

**Comment by delgermurun** [comment]
@arunoda Okay I'll. Thanks for suggestions & pressure :D.

> We will also need a test checking triggers as well.

Okay. I'm going to add more tests.

**Comment by delgermurun** [comment]
@arunoda updated :)

File: client/router.js [hunk 2]
```diff
@@ -302,6 +307,24 @@ Router.prototype._notfoundRoute = function(context) {
 Router.prototype.initialize = function() {
   var self = this;
   this._updateCallbacks();
+
+  // Implementing idempotent routing
+  // by overriding page.js`s "show" method.
+  // Why?
+  // It is impossible to bypass exit triggers,
```
Comment (arunoda) [inline client/router.js]: Okay. Now we are talking :)

**Comment by arunoda** [comment]
Great. This looks great for me.
I'll get this somewhere around today and do a release.

Good job on commenting :)

> Note: We should try to avoid comments as possible as we can by writing clean code with meaningful names. But, in cases like hacking some other module like this needs more reasoning.
> So, we need to do some comments on that.

**Comment by delgermurun** [comment]
@arunoda Thanks.

Regarding comment that makes a ton of sense :)

**Comment by arunoda** [comment]
Released with `1.15.0` :)

**Comment by delgermurun** [comment]
Nice.

**Pull request merged**